crates/sl-verifiable-enc/src/lib.rs

```diff
@@ -0,0 +1,651 @@
+// Copyright (c) Silence Laboratories Pte. Ltd. All Rights Reserved.
+// This software is licensed under the Silence Laboratories License Agreement.
+#[doc = include_str!("../README.md")]
+use ff::{Field, PrimeField};
+use group::{Group, GroupEncoding};
+use num_bigint_dig::ModInverse;
+use rand::{Rng, SeedableRng};
+use rand_chacha::{rand_core::CryptoRngCore, ChaCha20Rng};
+use rsa::{
+    traits::PublicKeyParts, BigUint, Pkcs1v15Encrypt, RsaPrivateKey,
+    RsaPublicKey,
+};
+use sha2::{Digest, Sha256};
+use std::ops::Index;
+use subtle::{Choice, ConditionallySelectable, ConstantTimeEq};
+use thiserror::Error;
+
+pub const SECURITY_PARAM: usize = 120;
+
+//Re-Exports
+pub use rsa;
+
+// Intentionally not giving too much information about the error
+#[derive(Debug, Error)]
+pub enum RsaError {
+    #[error("Error during encryption")]
+    EncError,
+    #[error("Error during decryption")]
+    DecError,
+    #[error("Invalid label inverse")]
+    InvalidLabel,
+    #[error("Verification failed")]
+    VerificationFailed,
+    #[error("Invalid SIZE parameter, must be equal to the size of the scalar in bytes")]
+    InvalidSizeParam,
+    #[error("(de)Serialization error")]
+    SerdeError(String),
+}
+
+pub struct ProofData<G: Group + GroupEncoding> {
+    g_r: G::Repr,
+    enc_x_r: Vec<u8>,
+    enc_r: Vec<u8>,
+}
+
+pub struct VerifiableRsaEncryption<G, const SIZE: usize>
+where
+    G: Group + GroupEncoding + ConstantTimeEq,
+    G::Scalar: ConditionallySelectable,
+{
+    pub seed: [u8; 32],
+    pub proofs: Vec<ProofData<G>>,
+    pub open_scalars: Vec<G::Scalar>,
+    security_param: usize,
+}
+
+impl<G, const SIZE: usize> VerifiableRsaEncryption<G, SIZE>
+where
+    G: Group + GroupEncoding + ConstantTimeEq,
+{
+    pub fn encrypt_with_proof<R: CryptoRngCore>(
+        x: &G::Scalar,
+        rsa_pubkey: &RsaPublicKey,
+        label: &[u8],
+        security_param: Option<usize>,
+        rng: &mut R,
+    ) -> Result<Self, RsaError>
+    where
+        <G::Scalar as PrimeField>::Repr: From<[u8; SIZE]>,
+    {
+        let seed = rng.gen::<[u8; 32]>();
+        let security_param = security_param.unwrap_or(SECURITY_PARAM);
+        let mut proofs = Vec::with_capacity(SECURITY_PARAM);
+        let q_point = G::generator() * x;
+        let mut r_list = Vec::with_capacity(SECURITY_PARAM);
```

Comment:
`Vec::with_capacity(SECURITY_PARAM);` -> `Vec::with_capacity(security_param);`

because we already figured out the correct value of `security_param`

Reply:
Oops forgot to update this one

Reply:
@leontiad @eclipse7 Should we not allow security_param below 120?